Preventing general instances from matching

I am working on adding a default instance for `Eq` that uses equality attributes, as mentioned in https://github.com/melt-umn/silver/issues/401:
```
destruct attribute compareTo;
equality attribute isEqual with compareTo;

instance attribute compareTo<a> occurs on a,
         attribute isEqual {compareTo} occurs on a
         => Eq a {
  eq = \ x::a y::a -> decorate x with {compareTo = y;}.isEqual;
}
```

Note that the above instance will initially match for decorated types.  This an issue because a) it leads to confusing error messages and b) we check if instances exist as a hint during type checking to help specialize possibly-decorated types.  For example an issue I found in Silver:
```
instance Eq TyVar {
  eq = \ tv1::TyVar tv2::TyVar -> tv1.kindrep == tv2.kindrep && tv1.extractTyVarRep == tv2.extractTyVarRep;
}

aspect production skolemType
top::Type ::= tv::TyVar
{
  top.unify = 
    case top.unifyWith of
    | skolemType(otv) when tv.kindrep == otv.kindrep ->
        if tv == otv then ... else ...
    | ...
    end;
}
```
Here `tv` and `otv` have possibly-decorated types that must be specialized for instance resolution.  Recall [the slight hack where we look up contexts when doing type inference for function application](https://github.com/melt-umn/silver/blob/ff76983b44bbf66f32346ebf7c48f5887e1c297c/grammars/silver/compiler/analysis/typechecking/core/Context.sv#L29): In the expression `tv == otv` (aka `eq(tv, otv)`), previously we would check for the existence of an instance for `Eq Decorated TyVar with i`, and not finding one specialize to just `TyVar` - the `Eq TyVar` instance then matches successfully.  

However the introduction of the default instance breaks this, as an instance for `Eq Decorated TyVar with i` is now found.  This means the specialization no longer happens, and instance resolution fails with a confusing error message when trying to satisfy the constraints on the default instance:
```
Unification.sv:13:11: error: Could not find an instance for attribute silver:core:isEqual {silver:core:compareTo} occurs on Decorated silver:compiler:definition:type:TyVar with {} (arising from the instance for silver:core:Eq Decorated silver:compiler:definition:type:TyVar with {}, arising from the use of silver:core:eq)
Unification.sv:13:11: error: Could not find an instance for attribute silver:core:compareTo<Decorated silver:compiler:definition:type:TyVar with {}> occurs on Decorated silver:compiler:definition:type:TyVar with {} (arising from the instance for silver:core:Eq Decorated silver:compiler:definition:type:TyVar with {}, arising from the use of silver:core:eq)
```

The problem here is that we really only want the default instance to match undecorated nonterminals, and not decorated types.  The fix I am considering is adding a way to explicitly cause instance matching to fail for a type:
```
delete instance Eq Decorated a with i;
```
or something, I could be convinced of better syntax.  

This feature could be more generally useful in helping avoid confusing errors caused by instance constraint resolution errors, by preventing specific types from matching a polymorphic instance with constraints in the first place.  Does anything like this exist in another functional language?

--- a/grammars/silver/compiler/analysis/typechecking/core/Context.sv
+++ b/grammars/silver/compiler/analysis/typechecking/core/Context.sv
@@ -32,7 +32,13 @@ top::Context ::= cls::String t::Type
     -- Only refine to the *undecorated* type based on instances, since if left
     -- unrefined the type will be treated as decorated, anyway (and leaving it
     -- unrefined will result in less confusing behavior.)
-    | ntOrDecType(nt, inhs, _) when !null(getInstanceDcl(cls, nt, top.env)) && null(getInstanceDcl(cls, decoratedType(nt, inhs), top.env)) ->
+    -- If there is an instance for the undecorated type, and all instances for
+    -- the decorated type are typeError instances (or there are none),
+    -- then specialize to the undecorated type, since the decorated type will
+    -- always give a type error.
+    | ntOrDecType(nt, inhs, _) when
+        !null(getInstanceDcl(cls, nt, top.env)) &&
+        all(map((.isTypeError), getInstanceDcl(cls, decoratedType(nt, inhs), top.env))) ->
       composeSubst(top.downSubst, substT.unifyInstanceNonterminal)
     | _ -> top.downSubst
     end;
@@ -118,6 +124,14 @@ top::Context ::= i1::Type i2::Type
       | (_, tvs1), (_, tvs2) when any(map(contains(_, tvs2), tvs1)) -> []
       | _, _ -> [err(top.contextLoc, s"${prettyTypeWith(i1, top.freeVariables)} is not a subset of ${prettyTypeWith(i2, top.freeVariables)} (arising from ${top.contextSource})")]
       end;
+
+  top.upSubst = top.downSubst; -- No effect on decoratedness
+}
+
+aspect production typeErrorContext
+top::Context ::= msg::String
+{
+  top.contextErrors := [err(top.contextLoc, msg ++ " (arising from " ++ top.contextSource ++ ")")];
 
   top.upSubst = top.downSubst; -- No effect on decoratedness
 }

--- a/grammars/silver/compiler/definition/env/Context.sv
+++ b/grammars/silver/compiler/definition/env/Context.sv
@@ -32,6 +32,9 @@ synthesized attribute contextClassName::Maybe<String> occurs on Context;
 
 synthesized attribute resolved::[DclInfo] occurs on Context;
 
+monoid attribute isTypeError::Boolean with false, || occurs on Contexts, Context;
+propagate isTypeError on Contexts, Context;
+
 aspect production instContext
 top::Context ::= cls::String t::Type
 {
@@ -154,12 +157,10 @@ top::Context ::= t::Type
 }
 
 synthesized attribute isTypeable::Boolean occurs on Type;
-aspect default production
-top::Type ::=
-{ top.isTypeable = true; }
-aspect production skolemType
-top::Type ::= _
-{ top.isTypeable = false; }
+aspect isTypeable on Type of
+| skolemType(_) -> false
+| _ -> true
+end;
 
 aspect production inhSubsetContext
 top::Context ::= i1::Type i2::Type
@@ -180,6 +181,20 @@ top::Context ::= i1::Type i2::Type
         !unifyDirectional(d.typeScheme.monoType, i1).failure && !d.typeScheme.monoType.isError &&
         !unifyDirectional(d.typerep2, i2).failure && !d.typerep2.isError,
       searchEnvTree("subset", top.env.instTree));
+}
+
+aspect production typeErrorContext
+top::Context ::= msg::String
+{
+  top.contextSuperDefs = \ DclInfo String Location -> [];
+  top.contextMemberDefs = \ [TyVar] String Location -> [];
+  top.contextSigDefs = \ NamedSignature String Location -> [];
+  top.contextSuperOccursDefs = \ d::DclInfo g::String l::Location -> [];
+  top.contextMemberOccursDefs = \ tvs::[TyVar] g::String l::Location -> [];
+  top.contextSigOccursDefs = \ ns::NamedSignature g::String l::Location -> [];
+  top.contextClassName = nothing();
+  top.resolved = [];
+  top.isTypeError <- true;
 }
 
 -- Invariant: This should be called when a and b are unifyable

--- a/grammars/silver/compiler/definition/env/DclInfo.sv
+++ b/grammars/silver/compiler/definition/env/DclInfo.sv
@@ -55,6 +55,7 @@ closed nonterminal DclInfo with sourceGrammar, sourceLocation, fullName, -- ever
                          classMembers, givenInstanceType, superContexts, typerep2, -- type classes, in the type namespace
                          namedSignature, hasForward, -- values that are fun/prod
                          attrOccurring, isAnnotation, -- occurs
+                         isTypeError, -- instances
                          isInherited, isSynthesized, -- attrs
                          prodDefs, -- production attributes
                          substitutedDclInfo, givenSubstitution -- type substitutions on dcls
@@ -94,6 +95,7 @@ top::DclInfo ::=
   top.superContexts = [];
 
   -- instances
+  top.isTypeError := false;
   top.typerep2 = error("Internal compiler error: must be defined for all binary constraint instances");
   
   -- Values that are not fun/prod have this valid default.
@@ -354,6 +356,8 @@ top::DclInfo ::= fn::String bound::[TyVar] contexts::[Context] ty::Type
   top.fullName = fn;
   
   top.typeScheme = constraintType(bound, contexts, ty);
+
+  top.isTypeError := any(map((.isTypeError), contexts));
 }
 abstract production instConstraintDcl
 top::DclInfo ::= fntc::String ty::Type tvs::[TyVar]

--- a/grammars/silver/compiler/definition/type/Type.sv
+++ b/grammars/silver/compiler/definition/type/Type.sv
@@ -77,6 +77,12 @@ top::Context ::= i1::Type i2::Type
   top.freeVariables = setUnionTyVars(i1.freeVariables, i2.freeVariables);
 }
 
+abstract production typeErrorContext
+top::Context ::= msg::String
+{
+  top.freeVariables = [];
+}
+
 {--
  - Silver Type Representations.
  -}

--- a/grammars/silver/compiler/definition/type/syntax/Constraint.sv
+++ b/grammars/silver/compiler/definition/type/syntax/Constraint.sv
@@ -221,6 +221,19 @@ top::Constraint ::= i1::TypeExpr 'subset' i2::TypeExpr
     end;
 }
 
+concrete production typeErrorConstraint
+top::Constraint ::= 'typeError' msg::String_t
+{
+  top.unparse = "typeError " ++ msg.lexeme;
+  top.contexts = [typeErrorContext(unescapeString(substring(1, length(msg.lexeme) - 1, msg.lexeme)))];
+
+  top.errors <-
+    case top.constraintPos of
+    | instancePos(_, _) -> []
+    | _ -> [err(top.location, "typeError constraint is only permitted on instances")]
+    end;
+}
+
 synthesized attribute classInstDcl::(DclInfo ::= String Type String Location);
 synthesized attribute occursInstDcl::(DclInfo ::= String Type Type String Location);
 synthesized attribute typeableInstDcl::(DclInfo ::= Type String Location);

--- a/grammars/silver/compiler/definition/type/syntax/Terminals.sv
+++ b/grammars/silver/compiler/definition/type/syntax/Terminals.sv
@@ -20,6 +20,7 @@ terminal InhSet_tkwd     'InhSet'     lexer classes {TYPE}; -- Well, actually a 
 
 terminal RuntimeTypeable_kwd 'runtimeTypeable' lexer classes {KEYWORD,RESERVED};
 terminal Subset_kwd          'subset'          lexer classes {KEYWORD};
+terminal TypeError_kwd       'typeError'       lexer classes {KEYWORD,RESERVED};
 
 -- Avoid making 'subset' reserved, for now
 disambiguate Subset_kwd, IdLower_t { pluck Subset_kwd; }

--- a/grammars/silver/compiler/definition/type/syntax/Terminals.sv
+++ b/grammars/silver/compiler/definition/type/syntax/Terminals.sv
@@ -20,7 +20,8 @@ terminal InhSet_tkwd     'InhSet'     lexer classes {TYPE}; -- Well, actually a 
 
 terminal RuntimeTypeable_kwd 'runtimeTypeable' lexer classes {KEYWORD,RESERVED};
 terminal Subset_kwd          'subset'          lexer classes {KEYWORD};
-terminal TypeError_kwd       'typeError'       lexer classes {KEYWORD,RESERVED};
+terminal TypeError_kwd       'typeError'       lexer classes {KEYWORD};
 
--- Avoid making 'subset' reserved, for now
+-- Avoid making these reserved, for now
 disambiguate Subset_kwd, IdLower_t { pluck Subset_kwd; }
+disambiguate TypeError_kwd, IdLower_t { pluck TypeError_kwd; }

--- a/grammars/silver/compiler/translation/java/core/InstanceDcl.sv
+++ b/grammars/silver/compiler/translation/java/core/InstanceDcl.sv
@@ -8,7 +8,7 @@ top::AGDcl ::= 'instance' cl::ConstraintList '=>' id::QNameType ty::TypeExpr '{'
   local contexts::Contexts = foldContexts(cl.contexts);
   contexts.boundVariables = boundVars;
 
-  top.genFiles := [pair(className ++ ".java", s"""
+  top.genFiles := if contexts.isTypeError then [] else [pair(className ++ ".java", s"""
 
 package ${makeName(top.grammarName)};
 

--- a/grammars/silver/compiler/translation/java/type/Context.sv
+++ b/grammars/silver/compiler/translation/java/type/Context.sv
@@ -136,6 +136,17 @@ top::Context ::= i1::Type i2::Type
 """;
 }
 
+aspect production typeErrorContext
+top::Context ::= msg::String
+{
+  -- This should never get translated
+  top.transType = error("Translation demanded for typeError context!");
+  top.transContext = error("Translation demanded for typeError context!");
+  top.transContextMemberName = error("Translation demanded for typeError context!");
+  top.transContextSuperAccessorName = error("Translation demanded for typeError context!");
+  top.transContextSuperAccessor = error("Translation demanded for typeError context!");
+}
+
 -- The translations of the narrowed forms of the instDcl's contexts are fed back as dcl.transContextDeps 
 inherited attribute transContextDeps::[String] occurs on DclInfo;
 attribute transContext occurs on DclInfo;

--- a/runtime/java/src/common/AppTypeRep.java
+++ b/runtime/java/src/common/AppTypeRep.java
@@ -61,6 +61,8 @@ public class AppTypeRep extends TypeRep {
 			}
 			String resultToString = it.hasNext()? it.next().toString() : "_";
 			result = "(" + resultToString + " ::= " + argsToString + ")";
+		} else {
+			result = a.toString();
 		}
 		if (it.hasNext()) {
 			result += "<" + it.next();

--- a/test/silver_features/TypeClasses.sv
+++ b/test/silver_features/TypeClasses.sv
@@ -423,3 +423,32 @@ wrongCode "i1 is not a subset of i2 (arising from the use of partialUndecorate)"
     return partialUndecorate(x);
   }
 }
+
+class NotDecThing a {
+  consumeNDT :: (String ::= a);
+}
+instance NotDecThing a {
+  consumeNDT = \ x::a -> reflectTypeName(x).fromJust;
+}
+instance typeError "Not decorated" => NotDecThing Decorated a with i {
+  consumeNDT = error("type error");
+}
+
+global ndt1::String = consumeNDT(pair(12, false));
+equalityTest(ndt1, "silver:core:Pair<Integer Boolean>", String, silver_tests);
+
+wrongCode "Not decorated (arising from the instance for silver_features:NotDecThing Decorated silver:core:Pair<Integer Boolean> with {}, arising from the use of consumeNDT)" {
+  global ndt2::String = consumeNDT(decorate pair(12, false) with {});
+}
+
+function ndtMaybeDec
+String ::= l::Location
+{
+  -- Should specialize to undecorated Location
+  return consumeNDT(l);
+}
+equalityTest(ndtMaybeDec(txtLoc("foo")), "silver:core:Location", String, silver_tests);
+
+wrongCode "typeError constraint is only permitted on instances" {
+  class typeError "bad" => TEClass a {}
+}

